### PR TITLE
Replace ocean/ice post NCL programs with Python program 

### PR DESCRIPTION
**Pull request by EricSinsky-NOAA**
**Description**

The ocean and ice post NCL programs can no longer be used in the global-workflow because NCL is obsolete. Therefore, the ocean and ice post NCL programs have been replaced by one Python program. The ocean/ice post Python program has the same capabilities as the NCL programs.

Refs #923 

**Type of change**

- [x] New feature (non-breaking change which adds functionality)

**How Has This Been Tested?**

- [x] EP4 on WCOSS2
  
**Checklist**

- [ ] My code follows the style guidelines of this project
- [x] I have performed a self-review of my own code
- [x] I have commented my code, particularly in hard-to-understand areas
- [ ] My changes need updates to the documentation. I have made corresponding changes to the documentation
- [x] My changes generate no new warnings
- [ ] New and existing tests pass with my changes
- [ ] Any dependent changes have been merged and published

### CODE REVIEW

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:142]: One cannot be compiling everytime this script is run.

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:145]: all imports must be at the top

**Comment by WalterKolczynski-NOAA** [comment]
@NeilBarton-NOAA @jiandewang Hold off on reviewing until all the checks passed and Rahul or I have given the go-ahead

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:170]: all new python programs must use `wxflow` to setup logging, execution etc.

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:184]: This needs to be handled with `argparse`
Use `raise` for exception handling and exit with an error message that is clear.

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:195]: Please provide comments or describe what the code/section is doing.
Break code into methods for readability and testing.
Add tests.

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py]: Where is this used?  Remove unused codes

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:268]: This should be turned into methods for testing and extensibility

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:277]: Can `model` be both `MOM` and `CICE` at the same time?

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:298]: Handle `MOM` in `MOM` related methods and `CICE` likewise.

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:1]: There are too many questions from this script.
1. Is this really efficient?  Is python the right choice for doing this or is a compiled language (e.g. Fortran) better suited for the operations of this size.
2. There are no tests for this.  Every python program that is added must include tests and pass standards.
3. Why are we using those `parm` files?  Can that information not be described in a better form e.g. `yaml` or `namelist` (if using Fortran).

**Review by aerorahul — CHANGES_REQUESTED** [review]
See first cut at the review comments.

File: ush/ocnpost.py
Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:142]: I agree. As of now, a compilation is only performed if the .so file does not exist in any Python path. However, the compile step can be removed completely from the code if that is preferred.

File: ush/ocnpost.py
Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:184]: Sure. I wasn't entirely sure if argparse was supported on WCOSS2. I will use argparse instead.

File: ush/ocnpost.py
Comment (aerorahul) [reply ush/ocnpost.py:142]: I don't understand why this can't just be this python function:
```python
def bilin(src, S, row, col, frac_b, dst_len):
    dst = [0.0] * dst_len  # Initialize destination field to 0.0

    # Apply weights
    for i in range(len(row)):
        dst[row[i]] += S[i] * src[col[i]]

    # Apply fraction correction
    for i in range(dst_len):
        if frac_b[i] != 0.0:
            dst[i] /= frac_b[i]

    return dst
```

File: ush/ocnpost.py
Comment (WalterKolczynski-NOAA) [reply ush/ocnpost.py:184]: argparse is core python

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:389]: these seem to be hard-wired here. Why?

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:423]: `netCDF4` affords lazy-loading of the dataset.  Why are we extracting that data upfront?  This doesn't seem like a good use of the library.

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:472]: these are things you would do in a Fortran code.
This is not proper use of the python language.

File: ush/ocnpost.py
Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:277]: This program can only process MOM and CICE one at a time, therefore `model` can only be either `MOM` or `CICE`.

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:446]: use `contextmanagers`

File: ush/ocnpost.py
Comment (aerorahul) [inline ush/ocnpost.py:498]: why are we writing out `NETCDF3` version filed?

File: ush/ocnpost.py
Comment (aerorahul) [reply ush/ocnpost.py:277]: In that case, this program needs to be written in a way that leverages OOP.

File: ush/ocnpost.py
Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:1]: 1.  I definitely agree that Fortran is more efficient with computation, which is why most of the intense computation is performed using a Fortran to Python interface generator (f2py) provided by Numpy. 
2. Sure, formal tests will be performed. 
3. There's no particular reason. I can change the format of the parm files to yaml.

File: ush/ocnpost.py
Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:389]: Yes, I will fix this so that the filenames not hard-wired. Thanks for catching this.

File: ush/ocnpost.py
Comment (EricSinsky-NOAA) [reply ush/ocnpost.py]: Thanks for catching this. I will remove this since it will not be needed.

File: ush/ocnpost.py
Comment (WalterKolczynski-NOAA) [reply ush/ocnpost.py:389]: All these things that are different for MOM and CICE should ideally be moved to a yaml file. The python shouldn't need to know what it is working on, just a yaml with all the settings.

File: ush/ocnpost.py
Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:142]: I tried something very similar to this. The program ran less efficiently when this was a python function.

File: ush/ocnpost.py
Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:389]: Agreed. yaml files will be used instead of the preliminary parm files that are being used now.

File: ush/ocnpost.py
Comment (aerorahul) [reply ush/ocnpost.py:142]: There are efficient bilinear interpolation methods in `numpy`.  Did you try any of those? 
If this is the meat and potatoes of this program, IMO, it makes sense for this to be a Fortran program.  The rest is just IO.

File: ush/ocnpost.py
Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:498]: I believe NETCDF3 was the format that was used in the ocean and ice post NCL programs. However, there is no particular reason we need to keep it this way. I'll change this to NETCDF4, or this can be a user-configurable option in a yaml file.

Comment (jiandewang) [reply ush/ocnpost.py:498]: > I believe NETCDF3 was the format that was used in the ocean and ice post NCL programs. However, there is no particular reason we need to keep it this way. I'll change this to NETCDF4, or this can be a user-configurable option in a yaml file.

it is netcdf4

File: ush/ocnpost.py
Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:142]: I tried [esmpy.Regrid](https://earthsystemmodeling.org/esmpy_doc/release/ESMF_8_0_1/html/regrid.html), but this also turned out to be less efficient.

File: ush/ocnpost.py
Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:498]: Got it, thanks Jiande.

File: ush/ocnpost.py
Comment (aerorahul) [reply ush/ocnpost.py:142]: I meant something from `numpy` or `scipy` e.g. https://numpy.org/devdocs/reference/generated/numpy.interp.html

Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:142]: To my understanding, `np.interp` does not have an option for a weighted bilinear interpolation. This Fortran subroutine, which uses the ESMF regridding weights in a bilinear or conservative interpolation is referenced [here](https://earthsystemmodeling.org/docs/release/ESMF_8_0_1/ESMF_refdoc/node3.html#SECTION03029300000000000000).

File: ush/ocnpost.py
Comment (HenryRWinterbottom) [reply ush/ocnpost.py:1]: 1. From my experience, FORTRAN is going to be the most efficient way to do this. Further, there is no need to wrap the FORTRAN with `f2py`. If you have the weights already, you should only need to loop over the coefficients to do the remapping. Using Python only complicates things and makes the application less portable.

2. A simple unit test would be using a static ESMF coefficients file, remapping from 5p0 to a comparable Gaussian grid nominal resolution and taking the difference of the fields. I would expect the different to be on the order of 0 aside from rounding error.

3. This is the perfect scenario for which to use a YAML configuration. The configuration file you provide is difficult to interpret. A YAML variation would be much more readable and easier to digest.

File: ush/ocnpost.py
Comment (HenryRWinterbottom) [reply ush/ocnpost.py:142]: I would personally avoid `numpy` and/or `scipy`. I have also tried these methods and they are slow and do not take into account issues such as landmask and fractional grids.

To reiterate, once you have the ESMF weights, all you need to do is add a short subroutine to read in the source grid variable and remap it to the destination grid. It should be short (on the order of 10 lines).

File: ush/ocnpost.py
Comment (HenryRWinterbottom) [reply ush/ocnpost.py:277]: The program should be agnostic to the model. All it requires if the ESMF coefficients and a FORTRAN bit to remap the variables. ESMF is also agnostic to the model. All it cares about is mapping points from a source grid to a destination grid. The only scenario that I can think of in which you would need to specify the model is if you are generating a MOM6 or CICE tri-polar projection from scratch.

File: ush/ocnpost.py
Comment (HenryRWinterbottom) [reply ush/ocnpost.py:298]: This is not necessary. Both MOM6 and CICE are on the same tri-polar grid for the UFS. Further, there is no rotation necessary for CICE since the variables are all Arakawa-C mass points.

File: ush/ocnpost.py
Comment (HenryRWinterbottom) [reply ush/ocnpost.py:498]: netCDF3 is an artifact of NGGPS.

File: ush/ocnpost.py
Comment (HenryRWinterbottom) [inline ush/ocnpost.py:1]: Anytime that you need to run over large loops in Python, consideration should be given to using a lower-level language (e.g., FORTRAN). Even with efficient use of Python list-comprehensions FORTRAN will still be a faster solution.

**Review by HenryRWinterbottom — COMMENTED** [review]
I think this application can be leaned out considerably.

1. Use YAML files as noted throughout the comments.

2. Take the remapping step out. If you already have the ESMF remapping coefficient files, all you need is a FORTRAN program to read the source variable files, do the interpolation, and write the remapped variable out.

3. The MOM6 velocity vector Earth -> grid and grid-> Earth relative rotations can also be done more efficiently in FORTRAN.

4. As is, this application limits portability.

File: ush/ocnpost.py
Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:142]: @HenryWinterbottom-NOAA Thank you for your feedback. Just to confirm, is this the short subroutine that you are referring to? 

           ! Initialize destination field to 0.0
           do i=1,dst_len
             dst(i)=0.0
           enddo

           ! Apply weights
           do i=1, n_s
             dst(row(i))=dst(row(i))+S(i)*src(col(i))
           enddo
           
           do i=1, dst_len
             if (frac_b(i) .ne. 0.0) then
               dst(i)=dst(i)/frac_b(i)
             endif
           enddo

Comment (HenryRWinterbottom) [reply ush/ocnpost.py:142]: @EricSinsky-NOAA Yes.

File: ush/ocnpost.py
Comment (JessicaMeixner-NOAA) [reply ush/ocnpost.py:298]: Can we confirm with @DeniseWorthen who wrote the original NCL scripts.  While I don't remember specific details, I do know that Denise is an expert on the CICE grid and carefully confirmed every variable was being interpolated the correct way and I'm assuming Eric is following what was done in the original NCL scripts here. Although maybe somethings changed since then. 

PS Thank you @EricSinsky-NOAA for this PR.  I know this has been a long standing need (https://github.com/NOAA-EMC/global-workflow/issues/923 was created a year ago) and it's greatly appreciated.

Comment (EricSinsky-NOAA) [reply ush/ocnpost.py:298]: Thanks @JessicaMeixner-NOAA. The vector rotation is being performed the same way as in Denise's NCL code. It is possible, however, that something has changed since then (as you pointed out).

**Pull request closed without merging**
